pkg/compact/compact_test.go

```diff
@@ -174,3 +178,275 @@ func BenchmarkGatherNoCompactionMarkFilter_Filter(b *testing.B) {
+				createBlockMeta(1, int64(time.Duration(2)*time.Hour/time.Millisecond), int64(time.Duration(4)*time.Hour/time.Millisecond), map[string]string{"b": "2"}, 0, []uint64{}),
+				createBlockMeta(2, int64(time.Duration(4)*time.Hour/time.Millisecond), int64(time.Duration(6)*time.Hour/time.Millisecond), map[string]string{"b": "2"}, 0, []uint64{}),
+				createBlockMeta(3, int64(time.Duration(6)*time.Hour/time.Millisecond), int64(time.Duration(8)*time.Hour/time.Millisecond), map[string]string{"a": "1", "b": "2"}, 1, []uint64{}),
+				createBlockMeta(4, int64(time.Duration(8)*time.Hour/time.Millisecond), int64(time.Duration(10)*time.Hour/time.Millisecond), map[string]string{"a": "1", "b": "2"}, 1, []uint64{}),
+			},
+			expected: map[string]planResult{
+				keys[0]: {
+					compactionRuns:   0.0,
+					compactionBlocks: 0.0,
+				},
+				keys[1]: {
+					compactionRuns:   0.0,
+					compactionBlocks: 0.0,
+				},
+				keys[2]: {
+					compactionRuns:   0.0,
+					compactionBlocks: 0.0,
+				},
+			},
+		},
+	} {
+		if ok := t.Run(tcase.testName, func(t *testing.T) {
+			blocks := make(map[ulid.ULID]*metadata.Meta, len(tcase.input))
+			for _, meta := range tcase.input {
+				blocks[meta.ULID] = meta
+			}
+			// form groups from the input metadata - do not hardcode groups. hence, grouper.Groups should stay
+			groups, err := grouper.Groups(blocks)
+			testutil.Ok(t, err)
+			ps := NewCompactionProgressCalculator(unRegisterer, planner)
+			err = ps.ProgressCalculate(context.Background(), groups)
+			testutil.Ok(t, err)
+			metrics := ps.CompactProgressMetrics
+			testutil.Ok(t, err)
+			for key := range tcase.expected {
+				a, err := metrics.NumberOfCompactionBlocks.GetMetricWithLabelValues(key)
+				testutil.Ok(t, err)
+				b, err := metrics.NumberOfCompactionRuns.GetMetricWithLabelValues(key)
+				testutil.Ok(t, err)
+				testutil.Equals(t, tcase.expected[key].compactionBlocks, promtestutil.ToFloat64(a))
+				testutil.Equals(t, tcase.expected[key].compactionRuns, promtestutil.ToFloat64(b))
+			}
+		}); !ok {
+			return
+		}
+	}
+}
+
+func TestDownsampleProgressCalculate(t *testing.T) {
+	reg := prometheus.NewRegistry()
+	unRegisterer := &receive.UnRegisterer{Registerer: reg}
+	logger := log.NewNopLogger()
+	type groupedResult map[string]float64
+
+	// pre calculating group keys
+	keys := make([]string, 3)
+	m := make([]metadata.Meta, 3)
+	m[0].Thanos.Labels = map[string]string{"a": "1"}
+	m[0].Thanos.Downsample.Resolution = downsample.ResLevel0
+	m[1].Thanos.Labels = map[string]string{"b": "2"}
+	m[1].Thanos.Downsample.Resolution = downsample.ResLevel1
+	m[2].Thanos.Labels = map[string]string{"a": "1", "b": "2"}
+	m[2].Thanos.Downsample.Resolution = downsample.ResLevel2
+	for ind, meta := range m {
+		keys[ind] = DefaultGroupKey(meta.Thanos)
+	}
+
+	var bkt objstore.Bucket
+	temp := promauto.NewCounter(prometheus.CounterOpts{Name: "test_metric_for_group", Help: "this is a test metric for downsample progress tests"})
```

Comment:
https://app.circleci.com/pipelines/github/thanos-io/thanos/7943/workflows/3f1a0011-dd6c-473e-acc0-bffafbd09f3e/jobs/17125.

When you use `promauto` library to register the metrics, please add `promauto.With(reg).NewCounter(...)` to specify the registry. Otherwise it will use the default global registry.

Reply:
I have done this in a recent commit but the CircleCI test is still failing.